### ISSUE TITLE
Some problem about batch normalize input sequence with mask.

### ISSUE THREAD

**Issue by ganji15** [erfannoury/seq2seq-lasagne#1]
I want use Batch Norm LSTM in my project. Fortunately, I found your impelemention of BNLSTM. However, I suspect there are some problems in your bathc normalization.

``` python

    # Layer Normalization
    def __ln__(self, z, alpha, beta):
        output = (z - z.mean(-1, keepdims=True)) / T.sqrt(z.var(-1, keepdims=True) + self._eps)
        output = alpha * output + beta
        return output
```

Whatif the input sequence have various length, which means the input mask should be provided. However,  I think your Layer Normalization doesn't consider the mask input. When calc means and vars of input sequences, we should ingore zeros inputs.

Here is some other layer normalization for sequence input with mask.

``` python
        #Sequence normalization

        # when calculating the mean and variance, only calculate among the 
        # frames that are not zero.

        total_sample = T.sum(mask)
        input_mean = input.sum(self.axes) / total_sample

        # need to be careful when calculating variance
        n_mask = mask.dimshuffle(1, 0, 'x')
        true_var = T.sqr(input - input_mean) * n_mask
        true_var = true_var.sum(self.axes) / total_sample
        input_inv_std = T.inv(T.sqrt(true_var + self.epsilon))
```

If my guess is right,  could you fix this problem? Thank you very much!

**Comment by erfannoury** [comment]
Hi,
What I have implemented here is Layer Normalization ([1607.06450](https://arxiv.org/abs/1607.06450)), not Batch Normalization. In Layer Normalization, each instance of a batch is normalized, independently from other instances. So there is no need to use mask for normalization. At least this is how I understood the paper. If you think it is wrong or if you have any more comments don't hesitate to ask.

**Comment by ganji15** [comment]
I do some experiments with you implementation of lnlstm, and it works very well!

However, I'm comfused by the Layer Normalization. It seems similar with Batch Normalization,but I cann't found out the difference between LN and BN. I mean the code of LN is  similar with BN.
 Could you explain the difference between them? 

**In Layer Normalization, each instance of a batch is normalized, independently from other instances.**
What does this mean? What's the difference with BN? 

``` python
    # Layer Normalization
    def __ln__(self, z, alpha, beta):
        output = (z - z.mean(-1, keepdims=True)) / T.sqrt(z.var(-1, keepdims=True) + self._eps)
        output = alpha * output + beta
        return output
```

If the input mask isn't considered,  how to correctly calculate the means and vars of sequences with variable time-step?

Thank you

**Comment by erfannoury** [comment]
In Batch Normalization, all training samples from a batch are used to compute a mean and variance value, and this mean and variance values are used to normalize the input to a layer during both training and test time. There is also the bias and gain values that are learned during training. However, this method of calculating the mean and variance is dependent on the batch size.
On the other hand, in Layer Normalization, a separate mean and variance value is calculated for each instance in a batch and these values are used to normalize each instance individually. Also gain and bias values in LN are learned during training (just like BN) and used during the testing phase. 
So to tell in another way, suppose in each batch you have `m` instances. In BN, a single mean and variance values are calculated for a batch whereas in LN `m` mean and `m` variance values are calculated and used to normalize the instances in a batch. I hope this makes things a little more clear.

BTW, it is good to take a look at [this](https://pdfs.semanticscholar.org/1827/de6fa9c9c1b3d647a9d707042e89cf94abf0.pdf?_ga=1.37283689.8546994.1456841600).

**Comment by ganji15** [comment]
I think you are right!

I read the paper "Layer Normalization" again. 
The paper says that "Unlike batch normalization, layer normaliztion does not impose any constraint on the size of a mini-batch."  The mean and variance values are related to the number of hiden units in a layer. 
So, the mean and variance values is related with hiden units in a recurrent network. However, every time step in a recurrent network shares the same hiden units. Is that right?

Thank you very much!

**Comment by erfannoury** [comment]
I think I didn't quite understand your question. But yes, every time step in a recurrent network shares the same set of hidden parameters. But it is not quite related to the Layer Normalization. In LN, each input instance is normalized and then consumed by the upcoming layer.

**Comment by ganji15** [comment]
# In origin paper:

## Layer Normalization

## ![question1](https://cloud.githubusercontent.com/assets/3733327/19428251/95095898-947b-11e6-9434-fd264fc6a0de.png)

### **According to the equation (3), the mean and variance value is calculated by previous sumed inputs ai,  and  H donates the number of hiden units, l donate the layer l, i donate the ith hiden unit index.**

### **So I think LN calculates mean and vars values in horizontal direction rather than vertical direction compared to BN. However , I doesn't understand that "different training cases have different normalization terms"**

---

# 

## LN in RNN

![question2](https://cloud.githubusercontent.com/assets/3733327/19428451/d626b87e-947c-11e6-9dc4-9b722c212b8e.png)

---

**Comment by erfannoury** [comment]
Sorry I have been busy these days and I haven't been able to reply. I'll re-read these papers and get back to you (hopefully) in less than 24 hours.

**Comment by ganji15** [comment]
It’s very kind of you to help me!

There is no hurry,  just reply me when you are free.

**Comment by erfannoury** [comment]
So I re-read the Batch Normalization paper and I think I am close to understanding difference between BN and LN. Suppose input to a layer are vectors `{x_i}` of size `n`, and `m` training cases in a mini-batch are given to a layer, so the input would be an `m-by-n` matrix. 
In BN, for feature `j` of input vectors `(x_i,j)`, a mean and a variance is estimated using all of the training cases in a mini-batch, i.e.

```
mu_j = (sum(x_i,j, over i) / m)
sigma_j = (sum((x_i,j - mu_j)^2, over i) / m)
```

After estimating these values, all features `j` for all training cases are normalized using `mu_j`, and `sigma_j`.
So we have 

```
xhat_i,j = (x_i,j - mu_j) / (sqrt(sigma_j))
```

In LN, each training case is normalized individually. So for training case `i`, we first estimate `mu_i` and `sigma_i`, i.e.

```
mu_i = (sum(x_i,j, over j) / n)
sigma_j = (sum(x_i,j - mu_i)^2, over j) / n)
```

After estimating these values for each training case, each of them is normalized accordingly, i.e.

```
xhat_i,j = (x_i,j - mu_i) / (sqrt(sigma_i))
```

Afterwards, each normalized value is scaled and added to a bias value, in both BN and LN. So we have

```
y_i,j = gamma_j * xhat_i,j + beta_j
```

**Comment by ganji15** [comment]
I think I have understanded BN and LN with your help!

Thank you very much!

**Comment by erfannoury** [comment]
It gave me a chance to read these papers once again. Happy to hear that it helped.
